pkg/webhook/kubernetes/warehouse/webhook.go

```diff
@@ -290,38 +273,73 @@ type subscriptionKey struct {
 
 type uniqueSubSet map[subscriptionKey]*field.Path
 
-func (s uniqueSubSet) addGit(sub kargoapi.GitSubscription, p *field.Path) error {
-	k := subscriptionKey{kind: "git", id: urls.NormalizeGit(sub.RepoURL)}
-	if _, exists := s[k]; exists {
-		return fmt.Errorf("subscription for Git repository already exists at %q", s[k])
-	}
-	s[k] = p
-	return nil
-}
-
-func (s uniqueSubSet) addImage(sub kargoapi.ImageSubscription, p *field.Path) error {
-	// The normalization of Helm chart repository URLs can also be used here
-	// to ensure the uniqueness of the image reference as it does the job of
-	// ensuring lower-casing, etc. without introducing unwanted side effects.
-	k := subscriptionKey{kind: "image", id: urls.NormalizeChart(sub.RepoURL)}
-	if _, exists := s[k]; exists {
-		return fmt.Errorf("subscription for image repository already exists at %q", s[k])
-	}
-	s[k] = p
-	return nil
-}
-
-func (s uniqueSubSet) addChart(sub kargoapi.ChartSubscription, isHTTP bool, p *field.Path) error {
-	k := subscriptionKey{kind: "chart", id: urls.NormalizeChart(sub.RepoURL)}
-	if isHTTP {
-		k.id = k.id + ":" + sub.Name
-	}
-	if _, exists := s[k]; exists {
+func (s uniqueSubSet) addSub(
+	f *field.Path,
+	sub kargoapi.RepoSubscription,
+) *field.Error {
+	// A small bit of special-casing is required here because, unlike generic
+	// subscriptions, the original three subscription types do not have one common
+	// way to identify them uniquely.
+	switch {
+	case sub.Chart != nil:
+		k := &subscriptionKey{
+			kind: "chart",
+			id:   urls.NormalizeChart(sub.Chart.RepoURL),
+		}
+		isHTTP := strings.HasPrefix(sub.Chart.RepoURL, "http://") || strings.HasPrefix(sub.Chart.RepoURL, "https://")
 		if isHTTP {
-			return fmt.Errorf("subscription for chart %q already exists at %q", sub.Name, s[k])
+			// For classical HTTP(S) Helm chart repositories, the chart name is part
+			// of the uniqueness criteria
+			k.id = k.id + ":" + sub.Chart.Name
+		}
+		if _, exists := s[*k]; exists {
+			var errMsg string
+			if isHTTP {
+				errMsg = fmt.Sprintf(
+					"subscription for chart %q already exists at %q",
+					sub.Chart.Name, s[*k],
+				)
+			} else {
+				errMsg = fmt.Sprintf("subscription for chart already exists at %q", s[*k])
+			}
+			return field.Invalid(f.Child("chart"), sub.Chart.RepoURL, errMsg)
+		}
+	case sub.Git != nil:
+		k := &subscriptionKey{
```

Comment:
Super nit: instead of getting a pointer to `subscriptionKey` and immediately dereferencing a few lines down, we should just use non-pointer values